### PR TITLE
Fix polymorphic mapping when some derived types have explicit mappings and others do not

### PR DESCRIPTION
**Pull request by kev-andrews**
as discussed in #4396

### CODE REVIEW

**Comment by kev-andrews** [comment]
@dotnet-policy-service agree

File: src/UnitTests/Projection/PolymorphismTests.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+﻿namespace AutoMapper.UnitTests.Projection;
+
+public class InMemoryMapPolymorphic : AutoMapperSpecBase
```
Comment (lbargaoanu) [inline src/UnitTests/Projection/PolymorphismTests.cs:3]: This needs to be an integration test in the dedicated folder.

File: src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs [hunk 2]
```diff
@@ -76,6 +76,10 @@ Expression CreateProjection(in ProjectionRequest request, LetPropertyMaps letPro
         foreach(var derivedMap in polymorphicMaps)
         {
             var sourceType = derivedMap.SourceType;
+            if (!request.SourceType.IsAssignableFrom(sourceType))
+            {
+                continue;
+            }
             var derivedRequest = request.InnerRequest(sourceType, derivedMap.DestinationType);
             var derivedProjection = CreateProjectionCore(derivedRequest, letPropertyMaps, derivedMap, Convert(source, sourceType));
```
Comment (lbargaoanu) [inline src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs:84]: I think the only change needed is to replace `Convert` with `TypeAs`.

**Review by lbargaoanu — CHANGES_REQUESTED** [review]

File: src/UnitTests/Projection/PolymorphismTests.cs (hunk 1 above)
Comment (kev-andrews) [reply src/UnitTests/Projection/PolymorphismTests.cs]: moved

File: src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs (hunk 2 above)
Comment (kev-andrews) [reply src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs:84]: you are correct, there is a difference in the generated Expressions and SQL though:

my initial solution does not add polymorphic maps that are not applicable to the expression tree:

```
DbSet<Car>()
    .Select(dtoCar => new VehicleModel{ Name = dtoCar.Name }
    )
    
SELECT [v].[Name]
FROM [Vehicles] AS [v]
WHERE [v].[Discriminator] = N'Car'
```

the approach of adding them with the type check results in:

```
DbSet<Car>()
    .Select(dtoCar => (dtoCar is Bicycle) ? new BicycleModel{ 
        EBike = (dtoCar as Bicycle).EBike, 
        Name = (dtoCar as Bicycle).Name 
    }
     : (dtoCar is Motorcycle) ? new MotorcycleModel{ 
        HasSidecar = (dtoCar as Motorcycle).HasSidecar, 
        Name = (dtoCar as Motorcycle).Name 
    }
     : new VehicleModel{ Name = dtoCar.Name }
    )

SELECT [v].[Id], [v].[Discriminator], [v].[Name], [v].[AmountDoors]
FROM [Vehicles] AS [v]
WHERE [v].[Discriminator] = N'Car'
```

if this is not an issue i can commit the version you proposed with less changes to the codebase

Comment (kev-andrews) [reply src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs]: ok, i cleaned up the commits to reflect the requested changes

**Review by lbargaoanu — APPROVED** [review]

**Pull request merged**